server/src/main/java/org/opensearch/action/bulk/TransportBulkAction.java

```diff
@@ -642,52 +652,62 @@ protected void doRun() {
                     bulkShardRequest::ramBytesUsed,
                     isOnlySystem
                 );
-                shardBulkAction.execute(bulkShardRequest, ActionListener.runBefore(new ActionListener<BulkShardResponse>() {
-                    @Override
-                    public void onResponse(BulkShardResponse bulkShardResponse) {
-                        for (BulkItemResponse bulkItemResponse : bulkShardResponse.getResponses()) {
-                            // we may have no response if item failed
-                            if (bulkItemResponse.getResponse() != null) {
-                                bulkItemResponse.getResponse().setShardInfo(bulkShardResponse.getShardInfo());
-                            }
 
-                            docStatusStats.inc(bulkItemResponse.status());
-                            responses.set(bulkItemResponse.getItemId(), bulkItemResponse);
-                        }
+                final Span span = tracer.startSpan(SpanBuilder.from("bulkShardAction", nodeId, bulkShardRequest));
+                try (SpanScope spanScope = tracer.withSpanInScope(span)) {
```

Comment:
If the onFailure/onResponse method of below listeners is never invoked, is it possible that is span may remain unended? Do we need to add a check on getting exception directly in this flow to ensure span is ended, at least for the flows exceptionally ending in the same thread?

Reply:
Added a catch block for ending spans to ensure spans are always closed